lib/strings: Add `sanitizeDerivationName` function

infinisil opened the pull request:
###### Motivation for this change
Was part of https://github.com/NixOS/nixpkgs/pull/78640, but put into its own PR after the discussion in https://github.com/NixOS/nixpkgs/pull/83223

Ping @Profpatsch @rpearce @Mic92 @roberth 

###### Things done

- [x] Manually verified that it does what it advertises
- [x] Added tests

infinisil committed:
lib/strings: Add sanitizeDerivationName function

Infinisil commented:
@rpearce I think the `safePrefix` thing in your PR can be achieved just by doing `validDerivationName "${safePrefix}-${string}"`

rpearce commented:
Would adding tests in something like https://github.com/NixOS/nixpkgs/blob/master/lib/tests/misc.nix be useful? I must confess that I don't know how to run them, though

rpearce commented:
:+1: works perfectly for me

roberth commented:
From the title of this PR I foolishly expected a `string -> bool`. May I suggest `sanitizeDerivationName`? It's accurate and it verbs.

Infinisil commented:
Changed the name to the suggested one and added some tests. Ideally the tests would generate random strings, put them through the function, then try to create a derivation with it. But our test infrastructure isn't really set up for that (yet?). So for now there's just a bunch of hardcoded ones.

roberth commented:
Some test improvements:
 - test the allowed characters by invoking `derivation`
 - test a string that is too long
 - test a string that is too long by itself, but not after bad chars are filtered out

Mic92 commented on lib/strings.nix:
```diff
@@ -678,4 +678,36 @@ rec {
+       sanitizeDerivationName pkgs.hello
+       => "-nix-store-2g75chlbpxlrqn15zlby2dfh8hr9qwbk-hello-2.10"
+  */
+  sanitizeDerivationName = lib.flip lib.pipe [
```
How much overhead do these compose function actually add?

Infinisil replied:
Shouldn't be too much. Though I have a change to it here which would improve performance a bit: https://github.com/NixOS/nixpkgs/pull/78640/commits/8022565db4a7f983ad9918990f9bd2e4a2a4a453

Profpatsch commented on lib/strings.nix:
```diff
@@ -678,4 +678,36 @@ rec {
+       sanitizeDerivationName pkgs.hello
+       => "-nix-store-2g75chlbpxlrqn15zlby2dfh8hr9qwbk-hello-2.10"
+  */
+  sanitizeDerivationName = lib.flip lib.pipe [
```
```suggestion
  sanitizeDerivationName = string: lib.pipe string [
```

rpearce replied:
(noob question about how this works!)

Does this remove the dependency on `lib.flip` by explicitly accepting the `string` as an argument before passing it on to `pipe`'s first argument? In essence, this sacrifices the point-free solution for 1 less dependency?

Infinisil replied:
It's less about reducing dependencies, but rather reducing mental overhead. By not using `flip` one doesn't need to know what `flip` does to understand this function.

Infinisil commented:
@roberth Now did that, and found 2 bugs in the function while doing so :D

One bug was that I used a regex like "[^+-.]", thinking it would mean `exclude any of "+", "-" and "."`, but really it meant `exclude the ascii range from "+" to "."`. So now I moved the "-" to the end of the brackets, so it's interpreted not as a range.

The other bug was regarding the derivation length: Nix also counts the ".drv" as being part of the name, so the name itself needs to be trimmed to 207 instead of 211 characters

@Profpatsch Also applied your suggestion

Infinisil commented:
One question to discuss: If the name is too long, should it be trimmed from the left or the right? Currently the function trims from the left, which would be better when names represent file names, where the left-most chars are the least specific. But trimming from the right would make more sense when it's a "${name}-${version}" like string, where the ${name} shouldn't be removed.

roberth commented on lib/tests/misc.nix:
```diff
@@ -478,4 +495,29 @@ runTests {
+  };
+
+  testSanitizeDerivationNameTooLongWithInvalid = testSanitizeDerivationName {
+    name = "&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&";
```
If you construct this by taking the longest valid name and adding an invalid char, you can test whether trimming and filtering happen in the right order. (this would be easy to tell from the impl if it wasn't for `lib.pipe`, but that's not the point)

Infinisil replied:
Done this.

roberth commented on lib/tests/misc.nix:
```diff
@@ -478,4 +495,29 @@ runTests {
+    expected = "-+--.-0123456789-=-?-ABCDEFGHIJKLMNOPQRSTUVWXYZ-_-abcdefghijklmnopqrstuvwxyz-";
+  };
+
+  testSanitizeDerivationNameTooLong = testSanitizeDerivationName {
```
This doesn't seem to get trimmed.

```suggestion
  testSanitizeDerivationNameLong = testSanitizeDerivationName {
```

Please also add a test case that is too long.

Infinisil replied:
Perhaps you meant the test case below this one? The one here does indeed get trimmed
